kuegi_bot/backtest_engine.py

```diff
@@ -170,45 +182,86 @@ def handle_order_execution(self, order: Order, intrabar: Bar):
+            else:
+                return order.limit_price + self.bars[0].close*shortFac
 
-    def handle_open_orders(self, intrabarToCheck: Bar) -> bool:
-        something_changed = False
+    def check_executions(self, intrabar_to_check: Bar, only_on_close):
         another_round= True
+        did_something= False
+        allowed_order_ids= None
+        if not only_on_close:
+            allowed_order_ids= set(map(lambda o:o.id, self.account.open_orders))
+        loopbreak= 0
         while another_round:
+            if loopbreak > 100:
+                print("got loop in backtest execution")
+                break
+            loopbreak += 1
             another_round= False
             should_execute= False
             for order in sorted(self.account.open_orders, key=self.orderKeyForSort):
+                if allowed_order_ids is not None and order.id not in allowed_order_ids:
+                    continue
+                force_taker= False
+                execute_order_only_on_close= only_on_close
+                if order.tstamp > intrabar_to_check.tstamp:
+                    execute_order_only_on_close= True # was changed during execution on this bar, might have changed the price. only execute if close triggered it
                 if order.limit_price is None and order.stop_price is None:
                     should_execute= True
                 elif order.stop_price and not order.stop_triggered:
-                    if (order.amount > 0 and order.stop_price < intrabarToCheck.high) or (
-                            order.amount < 0 and order.stop_price > intrabarToCheck.low):
+                    if (order.amount > 0 and order.stop_price < intrabar_to_check.high) or (
+                            order.amount < 0 and order.stop_price > intrabar_to_check.low):
                         order.stop_triggered = True
                         something_changed = True
                         if order.limit_price is None:
                             # execute stop market
                             should_execute= True
-                        elif ((order.amount > 0 and order.limit_price > intrabarToCheck.close) or (
-                                order.amount < 0 and order.limit_price < intrabarToCheck.close)):
+                            if only_on_close: # order just came in and executed right away: execution on the worst price cause can't assume anything better
+                                order.stop_price= intrabar_to_check.low if order.amount < 0 else intrabar_to_check.high
+
+                        elif ((order.amount > 0 and order.limit_price > intrabar_to_check.close) or (
+                                order.amount < 0 and order.limit_price < intrabar_to_check.close)):
                             # close below/above limit: got definitly executed
                             should_execute= True
+                            force_taker= True # need to assume taker.
```

Comment:
this was one of those changes, which makes backtest engine more reliable in sense of live executions?

Reply:
yes